skypy/halo/mass.py

```diff
@@ -1,63 +1,319 @@
+    rho_bar = (cosmology.critical_density0.to(units.Msun / units.Mpc ** 3)).value
+    rho_m0 = cosmology.Om0 * rho_bar
+
+    return rho_m0 * f_c * dlognu_dlogm / np.power(M, 2)
+
+
+def halo_mass_sampler(m_min, m_max, resolution, wavenumber, power_spectrum,
+                      growth_function, cosmology,
+                      collapse_function, params, size=None):
+    r'''Halo mass sampler.
+    This function samples haloes from their mass function,
+    see equation 7.46 in [1]_.
+
+    Parameters
+    -----------
+    m_min, m_max : array_like
+        Lower and upper bounds for the random variable m.
+    resolution: int
+        Resolution of the inverse transform sampling spline.
+    wavenumber : (nk,) array_like
+        Array of wavenumbers at which the power spectrum is evaluated,
+        in units of [Mpc^-1].
+    power_spectrum: (nk,) array_like
+        Linear power spectrum at redshift 0 in [Mpc^3].
+    growth_function : float
+        The growth function evaluated at a given redshift for the given
+        cosmology.
+    cosmology : astropy.cosmology.Cosmology
+        Cosmology object providing methods for the evolution history of
+        omega_matter and omega_lambda with redshift.
+    collapse_function: function
+        Collapse function to choose from a variety of models:
+        `sheth_tormen_collapse_function`, `press_schechter_collapse_function`.
+    params : tuple
+        List of parameters that determines the model used for
+        the halo mass function. The first parameter should be the name of
+        collapse function, followed by a second tuple with the parameters
+        of the chosen collapse model.
     size: int, optional
-         Output shape of luminosity samples.
-    x_min, x_max : float or int, optional
-        Lower and upper bounds in units of M*.
-    resolution : int, optional
-        Resolution of the inverse transform sampling spline. Default is 100.
+        Output shape of samples. Default is None.
 
     Returns
-    -------
-    mass : array_like
-        Drawn masses from the Press-Schechter mass function.
+    --------
+    sample: (size,) array_like
+        Samples drawn from the mass function, in units of solar masses.
 
     Examples
+    ---------
+    >>> import numpy as np
+    >>> from skypy.halo import mass
+    >>> from skypy.power_spectrum import eisenstein_hu
+
+    This example will sample from the halo mass function for
+    a Planck15 cosmology at redshift 0. The power spectrum is given
+    by the Eisenstein and Hu fitting formula:
+
+    >>> from astropy.cosmology import Planck15
+    >>> cosmo = Planck15
+    >>> D0 = 1.0
+    >>> k = np.logspace(-3, 1, num=100, base=10.0)
+    >>> A_s, n_s = 2.1982e-09, 0.969453
+    >>> Pk = eisenstein_hu(k, A_s, n_s, cosmo, kwmap=0.02, wiggle=True)
+
+    Sampling from the Sheth and Tormen mass function:
+
+    >>> mass.sheth_tormen(10**9, 10**12, 100, k, Pk, D0, cosmo)
+    29585260719.18624
```

Comment:
```suggestion
    >>> mass.sheth_tormen(10**9, 10**12, 100, k, Pk, D0, cosmo)
    29585260719.18...
```
This doctest is failing because the result is very slightly different between oldest and newest supported versions. Use an ellipsis so that the last few decimals are not checked.